### DISCUSSION TITLE
The media could not be loaded, either because the server or network failed or because the format is not supported.

### DISCUSSION THREAD

**Discussion by sisso** [roflcoopter/viseron#809] [Q&A]
Live cameras and motion recording are working and saving files. But I can not watch any recording on the browser. On the recording page, it display for each recording `The media could not be loaded, either because the server or network failed or because the format is not supported.`.

I tried with multiple computers, linux and windows, firefox and chrome. Downloading the recording locally I can play with any linux video player. 

I am using esp32cam to produce the images, do not appears to be a issue.

It is currently running on `Ubuntu 24.04.1 LTS (GNU/Linux 6.8.0-44-generic x86_64)` docker `roflcoopter/viseron:latest` 

Trying to force firefox accessing the file (using python http server on recording folder) it displays "No video with supported format and MIME type found", those are the headers "Content-type: video/mp4, Content-Length: 8481420"

From VCL video player. It describe as Encoded by `Lavf59.27.100` and the Codec: JPEG (jpg), Type: Video, Video resolution: 800x600, Buffer dimensions: 800x600, Decoded format: 24 bits RGB, Orientation: Top left, Color primarires: ITU-R BT.709, Colour transfer function: sRGB, Color psace: ITU-R Bt.709 Range

Enabling debug logs I have some warnings about not be able to find segment starts:

```
[2024-09-11 18:42:22] [DEBUG   ] [viseron.components.ffmpeg.recorder.cam4] - Segment information: {'20240911184211.mp4': {'start_time': 1726090931.0, 'end_time': 1726090935.96}, '20240911184206.mp4': {'start_time': 1726090926.0, 'end_time': 1726090931.0}, '20240911184216.mp4': {'start_time': 1726090936.0, 'end_time': 1726090941.0}}
[2024-09-11 18:42:22] [WARNING ] [viseron.components.ffmpeg.recorder.cam4] - Could not find matching start segment. Using earliest possible
[2024-09-11 18:42:22] [DEBUG   ] [viseron.components.ffmpeg.recorder.cam4] - Start event: 1726090922.821408, segment: 20240911184206.mp4
[2024-09-11 18:42:22] [DEBUG   ] [viseron.components.ffmpeg.recorder.cam4] - End event: 1726090937.815896, segment: 20240911184216.mp4
[2024-09-11 18:42:22] [DEBUG   ] [viseron.components.ffmpeg.recorder.cam4] - Concatenation command: ffmpeg -hide_banner -loglevel error -y -protocol_whitelist file,pipe -f concat -safe 0 -i - -c:v copy -c:a copy -movflags +faststart /recordings/cam4/2024-09-11/18:42:07.mp4
[2024-09-11 18:42:22] [DEBUG   ] [viseron.components.ffmpeg.recorder.cam4] - Segment script: 
file 'file:/segments/cam4/20240911184206.mp4'
file 'file:/segments/cam4/20240911184211.mp4'
file 'file:/segments/cam4/20240911184216.mp4'
outpoint 1
```

Any help is welcome

**Comment by roflcoopter** [comment]
What video and audio codecs is your camera using?

**Reply by sisso** [reply]
@roflcoopter  I didn´t expect this affect as it display normally and viseron is able to capture and save videos. But make sense, as I see nobody with this issue and I already try host in 2 different machines. 

I am using ESP32CAM.  

Connecting with VCL it display this:

```
Codec: Motion JPEG Video (MJPG)
Resolution: 800x600
Decoded format: Planar 4:2:2 YUV fullscale
Color space: ITU-R BT

```
mmpeg

```
ffmpeg -i http://192.168.1.2:81/stream -c copy -y my.mp4

Stream #0:0: Video: mjpeg (Baseline) (mp4v / 0x7634706D), yuvj422p(pc, bt470bg/unknown/unknown), 800x600, q=2-31, 25 tbr, 12800 tbn
```

**Comment by sisso** [comment]
Using a notebook camera as cam ip

`
docker run --device=/dev/video0:/dev/video0 -p56000:56000 -it gen2brain/cam2ip # on RPi use gen2brain/cam2ip:arm
`

I have same issue. I found those logs related with the "webcam"

```
[2024-09-23 14:20:19] [WARNING ] [viseron.components.ffmpeg.recorder.webcam] - Could not find matching start segment. Using earliest possible
[2024-09-23 14:21:13] [ERROR   ] [tornado.access] - 500 GET /api/v1/camera/webcam/snapshot?rand=xzqmn&width=520 (192.168.1.109) 1.76ms
[2024-09-23 14:21:28] [ERROR   ] [viseron.components.webserver.api.handlers] - Error in API CameraAPIHandler.get_snapshot: UUID('165bc8af-219f-448b-b32c-d97b780de92a')
Traceback (most recent call last):
  File "/src/viseron/components/webserver/api/handlers.py", line 290, in route_request
	getattr(self, route["method"])(*path_args, **path_kwargs)
  File "/src/viseron/components/webserver/api/v1/camera.py", line 141, in get_snapshot
	jpg = self._snapshot_from_memory(camera)
  File "/src/viseron/components/webserver/api/v1/camera.py", line 117, in _snapshot_from_memory
	ret, jpg = camera.get_snapshot(
  File "/src/viseron/domains/camera/init.py", line 474, in get_snapshot
	decoded_frame = self.shared_frames.get_decoded_frame_rgb(current_frame)
  File "/src/viseron/domains/camera/shared_frames.py", line 101, in get_decoded_frame_rgb
	return self._color_convert(shared_frame, COLOR_MODEL_RGB)
  File "/src/viseron/domains/camera/shared_frames.py", line 91, in _color_convert
	decoded_frame = self.get_decoded_frame(shared_frame).copy()
  File "/src/viseron/domains/camera/shared_frames.py", line 81, in get_decoded_frame
	return self._frames[shared_frame.name]
KeyError: UUID('165bc8af-219f-448b-b32c-d97b780de92a')
```

I probably want to manually steps whatever viseron is doing to capture the recordings and player with params. Just point me where I should start to look for would already help me out.

**Reply by roflcoopter** [reply]
That error is unrelated.

My guess is that it is because of the codec, most broswers only support h264 and some h265.
You would need to transcode the videos by setting `codec: h264` in your camera config under the `recorder` block

**Comment by sisso** [comment]
I changed all `cam.<id>recorder.codec` to h264.  Same issue for the esp32 cam, but, cam2ip, them webcan in the notebook is working. I added the `viseron_camera` in config example with external ip, that works too. 

So define the codec helps. But esp32cam are still have some peculiarity that cause things to fail. That is odd, as viseron can read, display live, detect changes and write the video.

**Reply by phanosp** [reply]
i think the problem is because the camera is sending on the stream that is being recorded a non-supported by your browser codec, i.e. h.265. In my case I had 2 identical camera when one was working but the other gave me this error. I used a software called cms and notice that in the camera that it was causing the issue the encoder settings where set to h.265. When I switch this yo h.264 any new recordings from that camera stared working again.

**Reply by sisso** [reply]
As far I search ESP32Cam (AI Thinker) is not capable to generate h265, only mjpeg... 

` mjpeg (Baseline), yuvj422p(pc, bt470bg/unknown/unknown)`

I am dumb on this, but I tried this with intention to read mjpeg and write h264 and didn´t solve the problem

```
$ ffmpeg -i http://192.168.1.3:81/stream -vcodec libx264 -y codec.mp4 -s 640x480

Input #0, mpjpeg, from 'http://192.168.1.3:81/stream':
  Duration: N/A, bitrate: N/A
  Stream #0:0: Video: mjpeg (Baseline), yuvj422p(pc, bt470bg/unknown/unknown), 640x480, 25 tbr, 25 tbn
Stream mapping:
  Stream #0:0 -> #0:0 (mjpeg (native) -> h264 (libx264))
```

There are something else

**Comment by sisso** [comment]
Some progress.....  I kept trying random ffmpeg arguments I found on internet and find one that is generating a valid video to run on local http server
````
$ ffmpeg -i http://192.168.1.3:81/stream -c:v libx264 -pix_fmt yuv420p test14-2.mp4
Input #0, mpjpeg, from 'http://192.168.1.3:81/stream':
  Duration: N/A, bitrate: N/A
  Stream #0:0: Video: mjpeg (Baseline), yuvj422p(pc, bt470bg/unknown/unknown), 640x480, 25 tbr, 25 tbn
Stream mapping:
  Stream #0:0 -> #0:0 (mjpeg (native) -> h264 (libx264))
````

Looks like the `-pix_fmt yuv420p` is the magic. 

I still need to try on viseron when I have a time.

**Comment by sisso** [comment]
Appears that the order that argument is used change the output.

This command line didn't work
```
$ ffmpeg -i http://192.168.1.3:81/stream -y -c:v libx264 test15-5.mp4 -pix_fmt yuv420p

Output #0, mp4, to 'test15-5.mp4':
  Metadata:
    encoder         : Lavf60.16.100
  Stream #0:0: Video: h264 (avc1 / 0x31637661), yuvj422p(pc, bt470bg/unknown/unknown, progressive), 640x480, q=2-31, 25 fps, 12800 tbn
    Metadata:
      encoder         : Lavc60.31.102 libx264
    Side data:
      cpb: bitrate max/min/avg: 0/0/0 buffer size: 0 vbv_delay: N/A
```

This works
```
$ ffmpeg -i http://192.168.1.3:81/stream -y -c:v libx264 -pix_fmt yuv420p test15-6.mp4

Output #0, mp4, to 'test15-6.mp4':
  Metadata:
    encoder         : Lavf60.16.100
  Stream #0:0: Video: h264 (avc1 / 0x31637661), yuv420p(tv, bt470bg/unknown/unknown, progressive), 640x480, q=2-31, 25 fps, 12800 tbn
    Metadata:
      encoder         : Lavc60.31.102 libx264
    Side data:
      cpb: bitrate max/min/avg: 0/0/0 buffer size: 0 vbv_delay: N/A
```

Notice on the first, the pix_fmt is ignored.

Viseron debug is printing this, where a lot of arguments are after the file name.

```
ffmpeg_cam1 -hide_banner -loglevel debug -avoid_negative_ts make_zero -fflags nobuffer -flags low_delay -strict experimental -fflags +genpts -use_wallclock_as_timestamps 1 -vsync 0 -timeout 5000000 -i http://192.168.1.5:81/stream -f segment -segment_time 5 -reset_timestamps 1 -strftime 1 -c:v copy /segments/cam1/%Y%m%d%H%M%S.mp4 -vf fps=1.0 -f rawvideo -pix_fmt yuv420p pipe:1
```

Trying to change this command and running locally with the output file on the end fail to even run

```
[vost#0:0/copy @ 0x5e5504a87b40] Filtergraph 'fps=1.0' was specified, but codec copy was selected. Filtering and streamcopy cannot be used together.
Error opening output file pipe:1.
Error opening output files: Function not implemented
```

I am testing on my `ffmpeg version 6.1.1-3ubuntu5`

I will keep investigating. Posting here just in case someone have any idea :)
